x/mongo/driver/topology/errors.go

```diff
@@ -62,16 +62,23 @@ func (e ServerSelectionError) Unwrap() error {
 
 // WaitQueueTimeoutError represents a timeout when requesting a connection from the pool
 type WaitQueueTimeoutError struct {
-	Wrapped error
+	Wrapped                      error
+	PinnedCursorConnections      uint64
+	PinnedTransactionConnections uint64
+	maxPoolSize                  uint64
 }
 
 // Error implements the error interface.
 func (w WaitQueueTimeoutError) Error() string {
 	errorMsg := "timed out while checking out a connection from connection pool"
 	if w.Wrapped != nil {
-		return fmt.Sprintf("%s: %s", errorMsg, w.Wrapped.Error())
+		errorMsg = fmt.Sprintf("%s: %s", errorMsg, w.Wrapped.Error())
 	}
-	return errorMsg
+
+	errorMsg = fmt.Sprintf("%s; maxPoolSize: %d, connections in use by cursors: %d, connections in use by transactions: %d",
+		errorMsg, w.maxPoolSize, w.PinnedCursorConnections, w.PinnedTransactionConnections)
+	return fmt.Sprintf("%s, connections in use by other operations: %d", errorMsg,
+		w.maxPoolSize-(w.PinnedCursorConnections+w.PinnedTransactionConnections))
```

Comment:
If I am correct about connections being pinned by both transactions and cursors, is it possible that this could result in a negative number? E.g. max pool size is 1, and the only connection is pinned by both.

Reply:
The `Connection#pin` function always increments the connection's reference count but only calls the provided `updatePoolFn` for the first reference (see the `if c.refCount == 0` conditional). So if a cursor is created in a transaction, the connection reference count should be 2, `pool.pinnedCursorConnections` should be 0, and `pool.pinnedTransactionConnections` should be 1. The math shouldn't yield a negative number in this case.